subgraphs/exchange/src/mapping_limitorder.ts

```diff
@@ -0,0 +1,227 @@
+  entity.fillReceipt_takerTokenFilledAmount = event.params.fillReceipt.takerTokenFilledAmount
+  entity.fillReceipt_remainingAmount = event.params.fillReceipt.remainingAmount
+  entity.fillReceipt_makerTokenFee = event.params.fillReceipt.makerTokenFee
+  entity.fillReceipt_takerTokenFee = event.params.fillReceipt.takerTokenFee
+  entity.takerTokenProfit = event.params.takerTokenProfit
+  entity.takerTokenProfitFee = event.params.takerTokenProfitFee
+  entity.takerTokenProfitBackToMaker = event.params.takerTokenProfitBackToMaker
+  entity.blockNumber = event.block.number
+  entity.blockTimestamp = event.block.timestamp
+  entity.transactionHash = event.transaction.hash
+  log.info('LimitOrderFilledByProtocol entity created at transaction hash: {}, this entity id: {}.', [event.transaction.hash.toHex(), entity.id])
+  // Entity must be saved after setting
+  entity.save()
+
+  // Record the tokens data of this LimitOrderFilledByProtocol event
+  addTradedToken(entity.fillReceipt_makerToken, event.block.timestamp)
+  addTradedToken(entity.fillReceipt_takerToken, event.block.timestamp)
+
+  // Get the order hash as order entity id
+  const orderId = event.params.orderHash.toHex()
+  // Load old or create new Order entity
+  let orderEntity = OrderEntity.load(orderId)
+  if (orderEntity == null) {
+    orderEntity = new OrderEntity(orderId)
+    orderEntity.orderStatus = 'Normal'
+    orderEntity.maker = event.params.maker
+    orderEntity.makerToken = event.params.fillReceipt.makerToken
+    orderEntity.takerToken = event.params.fillReceipt.takerToken
+    orderEntity.firstFilledTime = event.block.timestamp
+    orderEntity.lastFilledOrCancelledTime = event.block.timestamp
+    // Set the relationship of the first LimitOrderFilled entity under the same orderHash
+    const limitOrderFilledArray = new Array<string>(0)
+    limitOrderFilledArray.push(eventId)
+    orderEntity.limitOrderFilledId = limitOrderFilledArray
+    log.info('Order entity created at transaction hash: {}, this entity id: {}.', [event.transaction.hash.toHex(), orderEntity.id])
+    // If the order is filled all for the first time
+    if (event.params.fillReceipt.remainingAmount.equals(BigInt.fromI32(0))) {
+      orderEntity.orderStatus = 'FullyFilled'
+      log.info('Order is filled all, the order hash: {}.', [orderEntity.id])
+    }
+  } else {
+    // Set the relationship of this LimitOrderFilled entity under the same orderHash
+    const limitOrderFilledArray = orderEntity.limitOrderFilledId
+    limitOrderFilledArray.push(eventId)
+    orderEntity.limitOrderFilledId = limitOrderFilledArray
+    orderEntity.lastFilledOrCancelledTime = event.block.timestamp
+    // If the order is filled all for this time
+    if (event.params.fillReceipt.remainingAmount.equals(BigInt.fromI32(0))) {
+      orderEntity.orderStatus = 'FullyFilled'
+      log.info('Order is filled all, the order hash: {}.', [orderEntity.id])
+    }
+  }
```

Comment:
```suggestion
  let orderEntity = OrderEntity.load(orderId)
  if (orderEntity == null) {
    orderEntity = new OrderEntity(orderId)
    orderEntity.limitOrderFilledId = new Array<string>(0)
    orderEntity.orderStatus = 'Normal'
    orderEntity.maker = event.params.maker
    orderEntity.makerToken = event.params.fillReceipt.makerToken
    orderEntity.takerToken = event.params.fillReceipt.takerToken
    orderEntity.firstFilledTime = event.block.timestamp
  }
  // Set the relationship of this LimitOrderFilled entity under the same orderHash
  const limitOrderFilledArray = orderEntity.limitOrderFilledId
  limitOrderFilledArray.push(eventId)
  orderEntity.limitOrderFilledId = limitOrderFilledArray
  orderEntity.lastFilledOrCancelledTime = event.block.timestamp
  // If the order is filled all for this time
  if (event.params.fillReceipt.remainingAmount.equals(BigInt.fromI32(0))) {
    orderEntity.orderStatus = 'FullyFilled'
    log.info('Order is filled all, the order hash: {}.', [orderEntity.id])
  }
```

Reply:
I think the logic would be much clear with this code pattern, what do you think about it?

```typescript
// Find object
const obj = ...

// If object doesn't exist, set it up with default values
if (!obj) {
    obj.field = ...
}

// Then we could treat the obj as if it is always existing
if (obj.field === 0) {
    ...
}
```

Reply:
No problem! Also thank you for sharing its code pattern to me!